src/nvim/getchar.c

```diff
@@ -1970,7 +1970,7 @@ static int vgetorpeek(const int advance)
                 KeyNoremap = typebuf.tb_noremap[typebuf.tb_off];
                 del_typebuf(1, 0);
               }
-              exiting = !(c < 0 || (advance && c == NUL));
+              exiting = true;
```

Comment:
My argument against derived state holds true here. This is more error-prone and requires a non-trivial _proof of correctness_ described in your commit message.